### ISSUE TITLE
Test failure on 32-bit x86

### ISSUE THREAD

**Issue by jamesjer** [libsemigroups/libsemigroups#60]
Greetings, I maintain the GAP-related packages in the Fedora Linux distribution.  I just tried updating libsemigroups to version 0.6.0.  It built successfully on several architectures, but shows one test failure on 32-bit x86:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lt-lstest is a Catch v2.0.1 host application.
Run with -? for options
-------------------------------------------------------------------------------
RWS 22: F(2, 5); size 11, from Chapter 9, Section 1 in NR
-------------------------------------------------------------------------------
tests/rws.test.cc:392
...............................................................................
tests/rws.test.cc:403: FAILED:
  REQUIRE( rws.nr_rules() == 24 )
with expansion:
  49 == 24

0.000 s: RWS 22: F(2, 5); size 11, from Chapter 9, Section 1 in NR
===============================================================================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Let me know if I can provide more information to help debug this.

**Comment by james-d-mitchell** [comment]
Thanks for the report. Can you provide details of how you built libsemigroups in 32-bit mode, so that I can try and reproduce this?

**Comment by jamesjer** [comment]
Sure, you can see the attempted Fedora build here: https://koji.fedoraproject.org/koji/taskinfo?taskID=23661124.  All of the architectures except i686 succeeded (i686 build is here: https://kojipkgs.fedoraproject.org/work/tasks/1173/23661173/build.log).  Fedora provides a standard set of CFLAGS and LDFLAGS, which are passed to configure:

CFLAGS='-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m32 -march=i686 -fasynchronous-unwind-tables'
LDFLAGS='-Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld'

where those specs files contain the following.

 /usr/lib/rpm/redhat/redhat-hardened-cc1:
*cc1_options:
+ %{!r:%{!fpie:%{!fPIE:%{!fpic:%{!fPIC:%{!fno-pic:-fPIE}}}}}}

/usr/lib/rpm/redhat/redhat-annobin-cc1:
*cc1_options:
+ %{!-fno-use-annobin:%{!iplugindir*:%:find-plugindir()} -fplugin=annobin}

/usr/lib/rpm/redhat/redhat-hardened-ld:
*self_spec:
+ %{!static:%{!shared:%{!r:-pie}}}

*link:
+ -z now

This is with gcc 7.2.1 and binutils 2.29.1.  I can reproduce this on my local machine, so let me know if there is something I can do to help debug.

**Comment by james-d-mitchell** [comment]
Thanks.

I don't seem to be able to reproduce this, because I can't build libsemigroups with this setup. 

@jamesjer could you possibly try:

1) compiling in debug mode `./configure --enable-debug`

2) changing line 410 in `tests/rws.test.cc` to `rws.set_report(true);` and adding the line 
`std::cout << rws;` to the end of that `TEST_CASE`? 

and then post what this pukes out?

**Comment by jamesjer** [comment]
Sorry to go silent on you.  December got really busy.  I finally found time to wield a debugger today and I believe I found the problem.  I used gcc's sanitizers and was notified of a use-after-free situation.  That means the bug could manifest on any architecture, but apparently is more likely on i386, probably because it is such a resource-constrained architecture.  Line 496 of src/rws.cc does this:

            _stack.emplace(rule2);

But the stack is a stack of pointers, not objects, so emplace is the same as push; i.e., pointers to the same rule object are now on both _stack and _active_rules.  In the failing case, that rule is eventually moved to _inactive_rules, then deleted at the bottom of knuth_bendix().  Since a pointer to it is still in _stack, it is eventually dereferenced again.  Changing line 496 to this:

            _stack.push(new_rule(rule2));

appears to resolve the issue.  The gcc sanitizers no longer complain, and all tests pass on i386.

**Comment by james-d-mitchell** [comment]
@jamesjer thanks for this, I will  try to fix this in the next week or so.

**Comment by james-d-mitchell** [comment]
@jamesjer I'm trying to resolve this, but I still can't reproduce this, and I don't exactly follow what you wrote in your comment (sorry). 

You are correct that the following line puts the pointer `rule2` into the `_stack`
~~~~
_stack.emplace(rule2);
~~~~
but it is also removed from `_active_rules`  by the line (2 lines before `_stack.emplace(rule2)`)
~~~~
it = remove_rule(it)
~~~~
As an attempt to reproduce this, I added a substantial number of checks to make sure that there is no intersection between the rules pointed to by members of `_stack`, `_active_rules`, and `_inactive_rules`, and in all the tests every rule is where it should be. In particular, I checked that none of the rules being deleted at the end of `knuth_bendix` are in `_active_rules` and, indeed, they are not.  The `_stack` is empty by the end of `knuth_bendix`, so even if something were in both `_inactive_rules` and `_stack` it would no longer belong to `_stack` after `knuth_bendix` and so wouldn't eventually be dereferenced again. 

I'm also not sure I buy the limited resources argument, since this is a very small example, which shouldn't use anywhere close to the limit of the resources in a 32-bit build. For example, the maximum number of rules used in the RWS object is 45, and a rule is basically just a pair of strings, and the maximum length of these strings is at most 5, so probably the object uses a couple of kilobytes of storage at most. 

According to valgrind and ASAN there are no memory issues on my computer either (no use-after-free warnings), nor are there any valgrind warnings in the CI tests on travis. I noticed that one of the flags used to build `libsemigroups` is `-D_FORTIFY_SOURCE=2` and that this is not supported by ASAN according to 

https://github.com/google/sanitizers/issues/247

Could this be a false warning in this case with the issue elsewhere?

It'd be super helpful if you could provide me with the following:

* compile in debug mode: `./configure --enable-debug`

* changing line 410 in `tests/rws.test.cc` to `rws.set_report(true);` and adding the line
`std::cout << rws;` to the end of that `TEST_CASE`

and then post what this pukes out?

**Comment by jamesjer** [comment]
I apologize.  I omitted some important information.  The test failure is intermittent.  If I repeatedly run the test suite, I variously see (a) test RWS 22 fail as shown above (most frequently), (b) the entire suite complete successfully, (c) test RWS 51 fail in the same way, (d) test RWS 51 segfault, or (e) test RWS 64 fail similarly to above.  The intermittent nature of the failures led me to believe that a race condition might be involved, so I built for x86_64 with the thread sanitizer (which sadly is not available for i386).  A run of the tests with that in place issued the message about a use-after-free.  I cannot get the tests to fail under valgrind.  I've run them quite a few times and they always pass, and neither valgrind's helgrind tool nor its drd tool show any thread issues.  Also, even though the thread sanitizer warns about a use-after-free, the tests always pass on x86_64.

I also think I may have misread the use-after-free message.  It might be pointing at the strings inside the rule object, rather than the rule object itself.

I compiled in debug mode as requested.  The resulting data is only partly ASCII, and is rather large (nearly 2 MBytes).  I redirected it to a file, which is temporarily available here: [http://www.jamezone.org/libsemigroups.log](http://www.jamezone.org/libsemigroups.log).

Here is the thread sanitizer output, after building with only "-O2 -g -fsanitize=thread".  I note that it was issued by test Semigroup 26, not one of the tests that tend to fail on i386, so I could be barking up a bogus tree.

~~~~~~~~
==================
WARNING: ThreadSanitizer: data race (pid=2873)
  Write of size 8 at 0x7bcc0000c918 by thread T11:
    #0 std::_Bit_reference::operator=(bool) /usr/include/c++/7/bits/stl_bvector.h:87 (libsemigroups.so.0+0x00000004d5d7)
    #1 libsemigroups::Semigroup::idempotents(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&) src/semigroups.cc:1146 (libsemigroups.so.0+0x00000004d5d7)
    #2 void std::__invoke_impl<void, void (libsemigroups::Semigroup::*)(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&), libsemigroups::Semigroup*, unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&>(std::__invoke_memfun_deref, void (libsemigroups::Semigroup::*&&)(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&), libsemigroups::Semigroup*&&, unsigned long&&, unsigned long&&, unsigned long&&, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&) /usr/include/c++/7/bits/invoke.h:73 (libsemigroups.so.0+0x00000004ef11)
    #3 std::__invoke_result<void (libsemigroups::Semigroup::*)(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&), libsemigroups::Semigroup*, unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&>::type std::__invoke<void (libsemigroups::Semigroup::*)(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&), libsemigroups::Semigroup*, unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&>(void (libsemigroups::Semigroup::*&&)(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&), libsemigroups::Semigroup*&&, unsigned long&&, unsigned long&&, unsigned long&&, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&) /usr/include/c++/7/bits/invoke.h:95 (libsemigroups.so.0+0x00000004ef11)
    #4 decltype (__invoke((_S_declval<0ul>)(), (_S_declval<1ul>)(), (_S_declval<2ul>)(), (_S_declval<3ul>)(), (_S_declval<4ul>)(), (_S_declval<5ul>)())) std::thread::_Invoker<std::tuple<void (libsemigroups::Semigroup::*)(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&), libsemigroups::Semigroup*, unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul, 5ul>(std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul, 5ul>) /usr/include/c++/7/thread:234 (libsemigroups.so.0+0x00000004ef11)
    #5 std::thread::_Invoker<std::tuple<void (libsemigroups::Semigroup::*)(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&), libsemigroups::Semigroup*, unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&> >::operator()() /usr/include/c++/7/thread:243 (libsemigroups.so.0+0x00000004ef11)
    #6 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (libsemigroups::Semigroup::*)(unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&), libsemigroups::Semigroup*, unsigned long, unsigned long, unsigned long, std::vector<std::pair<libsemigroups::Element const*, unsigned long>, std::allocator<std::pair<libsemigroups::Element const*, unsigned long> > >&> > >::_M_run() /usr/include/c++/7/thread:186 (libsemigroups.so.0+0x00000004ef11)
    #7 <null> <null> (libstdc++.so.6+0x0000000bc42e)

  Previous read of size 8 at 0x7bcc0000c918 by thread T12:
    [failed to restore the stack]

  Location is heap block of size 102944 at 0x7bcc00000000 allocated by main thread:
    #0 operator new(unsigned long) <null> (libtsan.so.0+0x00000006fde6)
    #1 __gnu_cxx::new_allocator<unsigned long>::allocate(unsigned long, void const*) /usr/include/c++/7/ext/new_allocator.h:111 (lt-lstest+0x0000005210e7)
    #2 std::allocator_traits<std::allocator<unsigned long> >::allocate(std::allocator<unsigned long>&, unsigned long) /usr/include/c++/7/bits/alloc_traits.h:436 (lt-lstest+0x0000005210e7)
    #3 std::_Bvector_base<std::allocator<bool> >::_M_allocate(unsigned long) /usr/include/c++/7/bits/stl_bvector.h:492 (lt-lstest+0x0000005210e7)
    #4 std::vector<bool, std::allocator<bool> >::_M_fill_insert(std::_Bit_iterator, unsigned long, bool) /usr/include/c++/7/bits/vector.tcc:743 (lt-lstest+0x0000005210e7)
    #5 std::vector<bool, std::allocator<bool> >::insert(std::_Bit_const_iterator, unsigned long, bool const&) /usr/include/c++/7/bits/stl_bvector.h:995 (libsemigroups.so.0+0x00000004e260)
    #6 std::vector<bool, std::allocator<bool> >::resize(unsigned long, bool) /usr/include/c++/7/bits/stl_bvector.h:1036 (libsemigroups.so.0+0x00000004e260)
    #7 libsemigroups::Semigroup::init_idempotents() src/semigroups.cc:1012 (libsemigroups.so.0+0x00000004e260)
    #8 libsemigroups::Semigroup::cbegin_idempotents() tests/../src/semigroups.h:1162 (lt-lstest+0x00000061871a)
    #9 ____C_A_T_C_H____T_E_S_T____50 tests/semigroups.test.cc:1008 (lt-lstest+0x00000061871a)
    #10 Catch::TestInvokerAsFunction::invoke() const tests/catch.hpp:9025 (lt-lstest+0x00000041b6a9)
    #11 Catch::TestCase::invoke() const tests/catch.hpp:8926 (lt-lstest+0x0000004233b5)
    #12 Catch::RunContext::invokeActiveTestCase() tests/catch.hpp:7954 (lt-lstest+0x0000004233b5)
    #13 Catch::RunContext::runCurrentTest(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) tests/catch.hpp:7926 (lt-lstest+0x00000042f827)
    #14 Catch::RunContext::runTest(Catch::TestCase const&) tests/catch.hpp:7721 (lt-lstest+0x00000044539b)
    #15 runTests tests/catch.hpp:8162 (lt-lstest+0x000000445dfd)
    #16 Catch::Session::runInternal() tests/catch.hpp:8349 (lt-lstest+0x000000447bfe)
    #17 Catch::Session::run() tests/catch.hpp:8306 (lt-lstest+0x000000447d4d)
    #18 Catch::Session::run(int, char**) tests/catch.hpp:8274 (lt-lstest+0x000000447e3b)
    #19 main tests/catch.hpp:11716 (lt-lstest+0x00000040bb7f)

  Thread T11 (tid=2942, running) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x000000029483)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0x0000000bc714)
    #2 libsemigroups::Semigroup::cbegin_idempotents() tests/../src/semigroups.h:1162 (lt-lstest+0x00000061871a)
    #3 ____C_A_T_C_H____T_E_S_T____50 tests/semigroups.test.cc:1008 (lt-lstest+0x00000061871a)
    #4 Catch::TestInvokerAsFunction::invoke() const tests/catch.hpp:9025 (lt-lstest+0x00000041b6a9)
    #5 Catch::TestCase::invoke() const tests/catch.hpp:8926 (lt-lstest+0x0000004233b5)
    #6 Catch::RunContext::invokeActiveTestCase() tests/catch.hpp:7954 (lt-lstest+0x0000004233b5)
    #7 Catch::RunContext::runCurrentTest(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) tests/catch.hpp:7926 (lt-lstest+0x00000042f827)
    #8 Catch::RunContext::runTest(Catch::TestCase const&) tests/catch.hpp:7721 (lt-lstest+0x00000044539b)
    #9 runTests tests/catch.hpp:8162 (lt-lstest+0x000000445dfd)
    #10 Catch::Session::runInternal() tests/catch.hpp:8349 (lt-lstest+0x000000447bfe)
    #11 Catch::Session::run() tests/catch.hpp:8306 (lt-lstest+0x000000447d4d)
    #12 Catch::Session::run(int, char**) tests/catch.hpp:8274 (lt-lstest+0x000000447e3b)
    #13 main tests/catch.hpp:11716 (lt-lstest+0x00000040bb7f)

  Thread T12 (tid=2943, finished) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x000000029483)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0x0000000bc714)
    #2 libsemigroups::Semigroup::cbegin_idempotents() tests/../src/semigroups.h:1162 (lt-lstest+0x00000061871a)
    #3 ____C_A_T_C_H____T_E_S_T____50 tests/semigroups.test.cc:1008 (lt-lstest+0x00000061871a)
    #4 Catch::TestInvokerAsFunction::invoke() const tests/catch.hpp:9025 (lt-lstest+0x00000041b6a9)
    #5 Catch::TestCase::invoke() const tests/catch.hpp:8926 (lt-lstest+0x0000004233b5)
    #6 Catch::RunContext::invokeActiveTestCase() tests/catch.hpp:7954 (lt-lstest+0x0000004233b5)
    #7 Catch::RunContext::runCurrentTest(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) tests/catch.hpp:7926 (lt-lstest+0x00000042f827)
    #8 Catch::RunContext::runTest(Catch::TestCase const&) tests/catch.hpp:7721 (lt-lstest+0x00000044539b)
    #9 runTests tests/catch.hpp:8162 (lt-lstest+0x000000445dfd)
    #10 Catch::Session::runInternal() tests/catch.hpp:8349 (lt-lstest+0x000000447bfe)
    #11 Catch::Session::run() tests/catch.hpp:8306 (lt-lstest+0x000000447d4d)
    #12 Catch::Session::run(int, char**) tests/catch.hpp:8274 (lt-lstest+0x000000447e3b)
    #13 main tests/catch.hpp:11716 (lt-lstest+0x00000040bb7f)

SUMMARY: ThreadSanitizer: data race /usr/include/c++/7/bits/stl_bvector.h:87 in std::_Bit_reference::operator=(bool)
==================
1.700 s: Semigroup 26: cbegin_idempotents/cend, is_idempotent [2 threads]
~~~~~~~~

**Comment by james-d-mitchell** [comment]
Thanks for taking the time to do this @jamesjer, it is most helpful. 

I've fixed the data race pointed out by the sanitiser, this is caused by concurrently reading and writing to different positions in a `std::vector` of `bool`, which don't support this. 

In the log file you posted, there are 24 rules displayed, so this would seem to be the output of the test when it passed, rather than when it failed, right? Any chance you can post the output of a failed instance of the test? Also, just out of interest, why is the output only partly ASCII?

I'm still not sure how to resolve this, I'll stare hard at the code in the coming days, and see if that helps...

**Comment by flsmith** [comment]
Hi @jamesjer, I've been trying to reproduce this issue and haven't been able to.
Firstly, does this issue still exist? I notice that some more recent builds have passed (e.g. [here](https://koji.fedoraproject.org/koji/buildinfo?buildID=1016831))

I can't quite build the tests using your precise flags, but have with an extra ```-fPIC```. The tests then run fine for me (on Fedora 27, with binutils 2.29-6 and gcc 7.2.1)
Could you try building without the PIC/hardening-related flags and/or annobin? It would be useful to know if they trigger the issue.

**Comment by james-d-mitchell** [comment]
It's been nearly 2 years since the last comment on this issue, and we haven't at any point been able to reproduce the issue. So, I propose that we won't fix this issue directly, but only by general rewriting of the Knuth-Bendix implementation in libsemigroups.

**Comment by jamesjer** [comment]
I'm so sorry.  Clearly this fell off my radar again.  I've got my fingers in too many pies. :-(  The reason the Fedora builds started working @flsmith is that I patched the code as described in the Dec. 30 2017 comment, which made the test failure go away, but is actually wrong as the analysis by @james-d-mitchell shows.  I tried again today, and with gcc 9.2.1 and binutils 2.33.1, the test failure still occurs.  I tried stripping off the Fedora build flags, and got down to just -O2 -g, with the test failure still happening.  Then I dropped the optimization level down to -O, and the test succeeded.  That combined with the fact that the failure has only ever been observed on 32-bit x86 leads me to suspect that I'm seeing an i686-specific gcc optimization bug at work.  I will try to narrow it down and, if successful, will report it to the gcc developers.
